### DISCUSSION TITLE
Cycle through carousel

### DISCUSSION THREAD

**Discussion by CornyVanBeck** [AvaloniaUI/Avalonia#8218] [Q&A] [answered]
Hello everybody, 

is there a way to call `carousel.Next()` and if it is the last element it starts from the beginning like an endless loop?
How would one do this?
If I set `SelectedIndex = 0` the carousel will animate backwards to the first element.

Thanks in advance ;)

**Comment by maxkatz6** [comment]
> If I set SelectedIndex = 0 the carousel will animate backwards to the first element.

It probably would require special IPageTransition implementation https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Controls/Carousel.cs#L57

For instance https://docs.avaloniaui.net/docs/animations/pagetransitions

**Reply by CornyVanBeck** [reply]
Hey, thanks for the quick response. But I don't understand how i can achieve this with PageTransitions. I created a short screen recording to show my current situation that i want to solve.
![Peek 2022-05-29 11-39](https://user-images.githubusercontent.com/36736700/170861812-215acef1-c625-4cc0-aba4-55b098f9ae61.gif)

And instead scrolling back to the first picture i wanted that it continues with the first picture. Maybe I inherit from the Carousel Class and override somehow the behavior?

**Reply by maxkatz6** [reply]
Take a look at this method of CarouselPresenter. It gets "from" page and "to" page by index. And computes argument "forward" depending on the direction.  https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Controls/Presenters/CarouselPresenter.cs#L189

In your case I believe it should be enough to implement a custom page transition, and swap "from" and "to" arguments if "forward == false".

Otherwise, you would need to redefine some CarouselPresenter logic.

**Reply by timunie** [reply]
I have a similar presentation mode. I use a `TransitioningContentControl` for this which is bound to a `SelectedElement` in my `ViewModel`. I have implement `NextCommand` and `PreviousCommand` which will handle forward and backward for me. That way I was able to control the transition and behavior I needed. 

More: https://docs.avaloniaui.net/docs/controls/transitioningcontentcontrol 

It may not suit your needs, but I wanted to share my findings anyway. 

Happy coding
Tim

**Reply by CornyVanBeck** [reply]
> Take a look at this method of CarouselPresenter. It gets "from" page and "to" page by index. And computes argument "forward" depending on the direction. https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Controls/Presenters/CarouselPresenter.cs#L189
> 
> In your case I believe it should be enough to implement a custom page transition, and swap "from" and "to" arguments if "forward == false".
> 
> Otherwise, you would need to redefine some CarouselPresenter logic.

Thanks, I will try that out :)

**Comment by CornyVanBeck** [comment, marked as answer]
The Solution is even easier than expected.
I changed this line
https://github.com/AvaloniaUI/Avalonia/blob/dc72af60bf7de680822054a3bd8776ece4387d76/src/Avalonia.Base/Animation/PageSlide.cs#L97
to
`Value = -distance`
and this line
https://github.com/AvaloniaUI/Avalonia/blob/dc72af60bf7de680822054a3bd8776ece4387d76/src/Avalonia.Base/Animation/PageSlide.cs#L124
to
`Value = distance`

And now I have the expected behavior.

![Peek 2022-05-31 20-17](https://user-images.githubusercontent.com/36736700/171257316-676a0bb8-95d0-43da-9cfb-d8d7629200c2.gif)
